builder/builder.go

```diff
@@ -199,15 +204,142 @@ func (b *Builder) startRegistryProxy(ctx context.Context, caps apicaps.CapSet) (
+			if err != nil {
+				b.opt.Console.Warnf("Failed to stop registry proxy support container: %v", err)
+			}
+			return nil, false
+		}
+		b.opt.Console.VerbosePrintf("Started support container for registry proxy on port %d", darwinProxyPort)
+		addr = fmt.Sprintf("localhost:%d", darwinProxyPort)
+	}
+
 	b.opt.LocalRegistryAddr = addr
 
-	return func() {
+	closeFn := func() {
+		if isDarwin {
+			err := b.stopDarwinProxyHelper(ctx)
+			if err != nil {
+				b.opt.Console.Warnf("Failed to stop registry proxy support container: %v", err)
+			}
+		}
 		p.Close()
 		select {
 		case <-ctx.Done():
 		case <-doneCh:
 		}
-	}, true
+	}
+
+	return closeFn, true
+}
+
+const darwinProxyContainerName = "darwin-registry-proxy"
+
+// startDarwinProxyHelper: Since Docker Desktop (Mac) containers run in a VM, a
+// special host name, host.docker.internal, is made available to access the
+// host's network. Docker can only pull insecurely from localhost, so we use a
+// socat container to proxy localhost:<port> request back out to the local
+// registry proxy created above.
+func (b *Builder) startDarwinProxyHelper(ctx context.Context, registryPort int) (int, error) {
+	infoMap, err := b.opt.ContainerFrontend.ContainerInfo(ctx, darwinProxyContainerName)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to check info")
+	}
+
+	var exists bool
+	if info, ok := infoMap[darwinProxyContainerName]; ok && info.Status != "missing" {
+		exists = true
+	}
+
+	if exists {
+		err := b.opt.ContainerFrontend.ContainerRemove(ctx, true, darwinProxyContainerName)
+		if err != nil {
+			return 0, errors.Wrap(err, "failed to remove existing container")
+		}
+	}
+
+	runCfg := containerutil.ContainerRun{
+		NameOrID: darwinProxyContainerName,
+		ImageRef: "alpine/socat:1.7.4.4",
+		Ports: []containerutil.Port{
+			{
+				IP:            "127.0.0.1",
+				HostPort:      0, // Bind to available port
+				ContainerPort: 80,
+				Protocol:      containerutil.ProtocolTCP,
+			},
+		},
+		ContainerArgs: []string{
+			"tcp-listen:80,fork,reuseaddr",
+			fmt.Sprintf("tcp:host.docker.internal:%d", registryPort),
+		},
+	}
+
+	err = b.opt.ContainerFrontend.ContainerRun(ctx, runCfg)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to start support container")
+	}
+
+	infoMap, err = b.opt.ContainerFrontend.ContainerInfo(ctx, darwinProxyContainerName)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to check info")
+	}
+
+	info, ok := infoMap[darwinProxyContainerName]
+	if !ok {
+		return 0, errors.New("container does not exist")
+	}
+
+	if len(info.Ports) == 0 {
+		return 0, errors.New("container port not set")
+	}
+
+	parts := strings.Split(info.Ports[0], ":")
+	if len(parts) != 3 {
+		return 0, errors.Errorf("invalid container port format: %s", info.Ports[0])
+	}
+
+	assignedPort, err := strconv.Atoi(parts[1])
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to parse container port")
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
```

Comment:
this should be a config value, for cases where there's an overloaded system (like a macstadium instance), and the user wants a greater timeout value.